How-to for Figgy PostgreSQL migration (#29)
pulibrary/pulibrary.github.io: _posts/2023-11-29-figgy-migration-details.md

---
date: 2023-11-29 13:25:00 -0400
title: Figgy Migration Details
layout: default
---

> ## Setting up Logical Replication and Warm Standbys for PostgreSQL 15

Author: ```suggestion
## Migrating to PostgreSQL 15 with Logical Replication 
```

In our [previous post](2023-11-08-migrating-postgres-via-replication.md), we described how we decided to migrate a very large database using logical replication. In this post we will detail the steps we followed to make it happen.

### How to Set Up Logical Replication

Before the migration really began, we set up logical replication. This recreated our very large database in an upgraded version of PostgreSQL.

To set up logical replication, we started with the [AWS Database Migration Guide](https://docs.aws.amazon.com/dms/latest/sbs/chap-manageddatabases.postgresql-rds-postgresql-full-load-publisher.html), then expanded and adapted the steps to our local environment.

Here's how we did it:

First set up the publisher.
- Ensure the source system (the publisher) can accept requests from the target system (the subscriber): edit the pg_hba.conf file on the source system (for us it was our old PostgreSQL 10 machine):
``` /etc/postgresql/<version>/main/pg_hba.conf
host    all             all             <IP_of_subscriber/32>       md5
```
- Make the publisher keep detailed "write ahead logs" (WALs) to support logical replication:

> ``` /etc/postgresql/<version>/main/postgresql.conf

Reviewer: These filenames don't render in the preview. Should we put them somewhere else?

``` psql CLI
CREATE SUBSCRIPTION <project_name>_subscription
CONNECTION 'host=<publisher_IP> port=5432 dbname=<database_name> user=<postgres_admin_user> password=<postgres_admin_password>' PUBLICATION <project_name>_publication WITH copy_data=true;
```
Once everything is set up, wait for full replication to happen. You can validate that replication is complete by generating row counts in both databases and comparing those numbers:
``` psql CLI
WITH tbl AS
  (SELECT table_schema,
          TABLE_NAME
   FROM information_schema.tables
   WHERE TABLE_NAME not like 'pg_%'
     AND table_schema in ('public'))
SELECT table_schema,
       TABLE_NAME,
       (xpath('/row/c/text()', query_to_xml(format('select count(*) as c from %I.%I', table_schema, TABLE_NAME), FALSE, TRUE, '')))[1]::text::int AS rows_n
FROM tbl
ORDER BY rows_n DESC;
```

Once replication was complete, we wanted to be sure our upgraded database had resilience built in, so we set up a warm standby in addition to our usual backup process.

### How to Migrate the Application

You will want to modify the process below to suit your local environment. Make each step specific so you can copy and paste.

> Here's how we did it:

Author: ```suggestion
This is how we did it:
```

- Select a page to use for testing. Make sure it loads and save the URL.
- Open a terminal to each database server:
    * `ssh user@old-database`
    * `ssh user@new-database`
- Open a SQL terminal connected to `<database_name>` on both servers: `sudo -u postgres psql -d <database_name>`
- Check row counts for all tables between both servers:
```sql
WITH tbl AS
  (SELECT table_schema,
          TABLE_NAME
   FROM information_schema.tables
   WHERE TABLE_NAME not like 'pg_%'
     AND table_schema in ('public'))
SELECT table_schema,
       TABLE_NAME,
       (xpath('/row/c/text()', query_to_xml(format('select count(*) as c from %I.%I', table_schema, TABLE_NAME), FALSE, TRUE, '')))[1]::text::int AS rows_n
FROM tbl
ORDER BY rows_n DESC;
```
- Assuming they're close, move on. If they're far off, stop here. Something's broken.

> - Notify your users. We did this using Slack: `@here We're now starting scheduled maintenance of Figgy, it will be going into read only mode for up to 3 hours. We'll send another message here when it's ready for regular use again.`

Author: ```suggestion
- Notify your users. We did this using Slack: `@here We're now starting scheduled maintenance of Figgy, it will be going into read only mode for up to 3 hours. We'll send another message here when it's ready for regular use again.`
```

- Set your application to read-only mode (or shut it down), so nobody can add new records while you migrate.
- Check row counts for all tables between both servers again (belt and suspenders, folks).
- If they're not exact, wait a minute, try again. If they never become exact, then stop. Otherwise, continue.

> - [Update the sequences](https://wiki.postgresql.org/wiki/Fixing_Sequences) in the new database.

Reviewer: ```suggestion
```

> - Run the following query in the new database's psql terminal:

Reviewer: We give the link but then copy its contents below. I suggest just having one line like
```suggestion
- Run the following query (which we pulled from the [postgresl.org wiki](https://wiki.postgresql.org/wiki/Fixing_Sequences)) in the new database's psql terminal to update the sequences:
```

> - Run the following query in the new database's psql terminal.

Author: ```suggestion
- Run the following query in the new database's psql terminal:
```

```sql
SELECT 
    'SELECT SETVAL(' ||
       quote_literal(quote_ident(sequence_namespace.nspname) || '.' || quote_ident(class_sequence.relname)) ||
       ', COALESCE(MAX(' ||quote_ident(pg_attribute.attname)|| '), 1) ) FROM ' ||
       quote_ident(table_namespace.nspname)|| '.'||quote_ident(class_table.relname)|| ';'
FROM pg_depend 
    INNER JOIN pg_class AS class_sequence
        ON class_sequence.oid = pg_depend.objid 
            AND class_sequence.relkind = 'S'
    INNER JOIN pg_class AS class_table
        ON class_table.oid = pg_depend.refobjid
    INNER JOIN pg_attribute 
        ON pg_attribute.attrelid = class_table.oid
            AND pg_depend.refobjsubid = pg_attribute.attnum
    INNER JOIN pg_namespace as table_namespace
        ON table_namespace.oid = class_table.relnamespace
    INNER JOIN pg_namespace AS sequence_namespace
        ON sequence_namespace.oid = class_sequence.relnamespace
ORDER BY sequence_namespace.nspname, class_sequence.relname;
```

> - this will output a new query which you should copy and paste and run.

Author: ```suggestion
- This will output a new query which you should copy and paste and run.
```

- Update the application configuration to point to the new database.
- Restart your web server and any workers.
- Make sure your test page loads.

> - We have multiple servers in a load-balanced setup, so we did those three steps multiple times. We removed half of our servers from the load balancer, updated them, reinstated them, removed the other half from the load balancer, restarted the web servers and workers, then checked that our test object still loaded before repeating those steps on our other servers. This gave us a way to retreat if we needed to. It also kept the system available for read operations during the upgrade.

Reviewer: "those three steps" is sort of confusing since the steps aren't numbered. What if we say "we updated the application server configuration in two batches"
```suggestion
- We have multiple servers in a load-balanced setup, so we updated the application server configuration in two batches. We removed half of our servers from the load balancer, updated them, reinstated them, removed the other half from the load balancer, restarted the web servers and workers, then checked that our test object still loaded before repeating those steps on our other servers. This gave us a way to retreat if we needed to. It also kept the system available for read operations during the upgrade.
```

- Take the application out of read-only mode, or start it up again. Do not advertise this yet.
- Try any other operations your application offers, make sure they work.
- Check your monitoring systems, make sure they don't show more errors than normal.
- If good, throw a party.
- If bad, put it back in read only mode, check the time, and either resolve or undo everything.
- Turn off logical replication by deleting the subscription on your new database (`DROP SUBSCRIPTION <project_name>_subscription;`)

> - Set up and enable your preferred backup strategy for your new database, verify.

Reviewer: Maybe swap this one below the user notification step
